### DISCUSSION TITLE
Duckdb python library disk spilling in aws lambda

### DISCUSSION THREAD

**Discussion by alahav123** [duckdb/duckdb#14976] [Q&A]
Here's a more polished and concise version of your description:

---

**Issue:**

I'm using **DuckDB** inside an **AWS Lambda** function, which has 10GB of RAM and 10GB of disk space. The issue arises when running a query, where **DuckDB** fails to spill data to disk, leading to a **memory limit exceeded** error (AWS Lambda error). The query I'm executing is as follows:

```sql
SELECT
    event_name,
    col_2,
    COUNT(DISTINCT event_id) AS value
FROM events
GROUP BY event_name, col_2
```

### DuckDB Connection Configuration:
I have configured the **DuckDB** connection as follows to try and limit memory usage and enable disk spilling:

```python
duckdb.connect(
    database=':memory:', 
    config={
        'memory_limit': '3GB',
        'max_temp_directory_size': '9GB',
        'temp_directory': '/tmp',
        'max_memory': '3GB',
        'preserve_insertion_order': False,
    }
)
```

### Problem:

Despite these settings, DuckDB is not limiting its memory usage and is not spilling to disk. The query fails with a **memory limit exceeded** error when processing large data sets, even though **AWS Lambda** has a 10GB memory and disk limit.

I have tested this issue in a **Docker environment** (using the image below), but the behavior is the same:

```dockerfile
FROM --platform=linux/amd64 public.ecr.aws/lambda/python:3.8

COPY requirements.txt .
RUN pip3 install -r requirements.txt
COPY *.py ${LAMBDA_TASK_ROOT}

ENV HOME=/tmp

CMD ["main.event_handler"]
```

### Request:

Could you suggest what might be going wrong with the **DuckDB memory management** and why it's not spilling data to disk as expected? Any guidance on resolving this memory issue in the Lambda environment would be greatly appreciated.

---

This version is clearer, more concise, and emphasizes the core issue and context, which should help others understand and troubleshoot the problem more effectively.

**Comment by szarnyasg** [comment]
Hi @alahav123, great question.

There are some limitations around the memory limit, namely that it only affects the memory handled by the buffer manager.

See here: https://duckdb.org/docs/configuration/pragmas#memory-limit

> The specified memory limit is only applied to the buffer manager. For most queries, the buffer manager handles the majority of the data processed. However, certain in-memory data structures such as [vectors](https://duckdb.org/docs/internals/vector.html) and query results are allocated outside of the buffer manager. Additionally, [aggregate functions](https://duckdb.org/docs/sql/functions/aggregates.html) with complex state (e.g., list, mode, quantile, string_agg, and approx functions) use memory outside of the buffer manager. Therefore, the actual memory consumption can be higher than the specified memory limit.

If you have complex aggregate functions, they may use memory outside of the buffer manager.

If not, it would be interesting to try to figure out whether the error is caused by containerization, the lambda environment, the Python environment, etc. To this end, it would be great if you could run the Python script locally on your computer and observe whether it starts spilling to disk. If not, then it's a more generic DuckDB issue. If it is spilling to disk, then the issue lies in (one of the) environments.
